### PR TITLE
fix: Install generator to location start with tild ~ 

### PR DESCRIPTION
**Pull request by tomer-epstein**

### CODE REVIEW

**Comment by tomer-epstein** [comment]
> I see several problems:
> 
> 1. ~ comes from linux/unix, does not exist in windows
> 2. ~~ would give wrong result

1. Right. In windows ithe tild will be replace with home dir.
2.  Sure. Any wrong path would give wrong result. It replace only the 1st tild. I see this as good solution, any ither tild us oart of the path.

**Review by slavik-lvovsky — COMMENTED** [review]
This , and other similar, setting property were designed to support only full path that user should provide. A path should be valid because today we do not provide validation for it.
We should probably improve the description of the property but not support all available linux path short hands. Especially when this property is used also in windows.

**Comment by tomer-epstein** [comment]
> This , and other similar, setting property were designed to support only full path that user should provide. A path should be valid because today we do not provide validation for it.
> We should probably improve the description of the property but not support all available linux path short hands. Especially when this property is used also in windows.

What do u suggest? To drope this fix?
I dont think it is posdible to validate textfield value in the prefrence.

**Comment by slavik-lvovsky** [comment]
> > This , and other similar, setting property were designed to support only full path that user should provide. A path should be valid because today we do not provide validation for it.
> > We should probably improve the description of the property but not support all available linux path short hands. Especially when this property is used also in windows.
> 
> What do u suggest? To drop this fix?
> I dont think it is posdible to validate textfield value in the prefrence.

Need to validate a provided path in exploregens.ts. In case that a path does not exist or invalid we should use the npm public installation path.
The description of the property should explain that the installation path must be valid and absolute.

File: backend/src/exploregens.ts [hunk 1]
```diff
@@ -18,7 +19,8 @@ export enum GenState {
 
 export class ExploreGens {
     public static getInstallationLocation(wsConfig: any) {
-        return _.trim(wsConfig.get(ExploreGens.INSTALLATION_LOCATION));
+        let location = _.trim(wsConfig.get(ExploreGens.INSTALLATION_LOCATION));
```
Comment (slavik-lvovsky) [inline backend/src/exploregens.ts:22]: it should be const

File: backend/tests/exploregens.spec.ts [hunk 2]
```diff
@@ -60,6 +61,7 @@ import { fail } from "assert";
 describe('exploregens unit test', () => {
     let sandbox: any;
```
Comment (slavik-lvovsky) [inline backend/tests/exploregens.spec.ts:62]: type of sandbox should be SinonSandbox
type of every mock should be SinonMock

File: backend/package.json [hunk 3]
```diff
@@ -135,7 +135,7 @@
 				},
 				"ApplicationWizard.installationLocation": {
 					"type": "string",
-					"markdownDescription": "Install generators in a specified location. If not set, global location is used."
+					"markdownDescription": "Set the default folder in which generators will be installed. The path to this folder must be valid and absolute. If there is no path defined, the global location will be used."
```
Comment (slavik-lvovsky) [inline backend/package.json:138]: With this solution we need to mention that the path must exists

File: backend/tests/exploregens.spec.ts [hunk 4]
```diff
@@ -324,16 +329,27 @@ describe('exploregens unit test', () => {
...
         it("location is a string with unnecessary spaces", () => {
-            workspaceConfigMock.expects("get").withExactArgs(ExploreGens["INSTALLATION_LOCATION"]).returns(`   ${TESTVALUE}   `);
+            const customLocation = `   ${TESTVALUE}   `;
+            workspaceConfigMock.expects("get").withExactArgs(ExploreGens["INSTALLATION_LOCATION"]).returns(customLocation);
+            fsMock.expects("existsSync").withExactArgs(TESTVALUE).returns(true);
             const res = exploregens["getGeneratorsLocationParams"]();
             expect(res).to.be.deep.equal(`--prefix ${TESTVALUE}`);
         });
+
+        it("location starts with tild ~", () => {
```
Comment (slavik-lvovsky) [inline backend/tests/exploregens.spec.ts:347]: this test should have fsMock.expects("existsSync").withExactArgs(TESTVALUE).returns(false);

**Review by slavik-lvovsky — CHANGES_REQUESTED** [review]

File: backend/tests/exploregens.spec.ts (hunk 2 above)
Comment (tomer-epstein) [reply backend/tests/exploregens.spec.ts:62]: it's the same in all our spec files!
https://github.com/SAP/yeoman-ui/blob/DEVXBUGS-8621/backend/tests/yeomanui.spec.ts#L39

File: backend/package.json (hunk 3 above)
Comment (tomer-epstein) [reply backend/package.json:138]: approved by Paola.
must be valid --> means it's exist. (valid)

File: backend/tests/exploregens.spec.ts (hunk 4 above)
Comment (tomer-epstein) [reply backend/tests/exploregens.spec.ts:347]: **Better** not mock it. location is invalid so existSync will return false.

File: backend/tests/exploregens.spec.ts (hunk 2 above)
Comment (slavik-lvovsky) [reply backend/tests/exploregens.spec.ts:62]: Right, it was like this until Alex Gilin found the best option. We should gradually replace all any by right type

File: backend/tests/exploregens.spec.ts (hunk 4 above)
Comment (slavik-lvovsky) [reply backend/tests/exploregens.spec.ts:347]: In test environment accidentally this path may exist and the test will fail. It is better to control it with mock

Comment (tomer-epstein) [reply backend/tests/exploregens.spec.ts:347]: 0.01% that accidentally contains a path equal to '~/notExistLocation'

Comment (slavik-lvovsky) [reply backend/tests/exploregens.spec.ts:347]: It is always better 100%, especially for tests.

File: backend/tests/exploregens.spec.ts [hunk 5]
```diff
@@ -353,7 +369,7 @@ describe('exploregens unit test', () => {
         });
 
         it("recommended array is undefined", () => {
-            workspaceConfigMock.expects("get").withExactArgs(exploregens["SEARCH_QUERY"]).returns();
+            workspaceConfigMock.expects("get").withExactArgs(exploregens["SEARCH_QUERY"]).returns("");
```
Comment (slavik-lvovsky) [inline backend/tests/exploregens.spec.ts:372]: it is better to remove .returns("") or .returns() if you do not use the returned value

File: backend/tests/exploregens.spec.ts [hunk 6]
```diff
@@ -318,22 +323,33 @@ describe('exploregens unit test', () => {
         });
 
         it("location undefined", () => {
-            workspaceConfigMock.expects("get").withExactArgs(ExploreGens["INSTALLATION_LOCATION"]).returns();
+            workspaceConfigMock.expects("get").withExactArgs(ExploreGens["INSTALLATION_LOCATION"]).returns("");
```
Comment (slavik-lvovsky) [inline backend/tests/exploregens.spec.ts:326]: it is better to remove .returns("") or .returns() if you do not use the returned value

File: backend/tests/exploregens.spec.ts [hunk 7]
```diff
@@ -214,7 +219,7 @@ describe('exploregens unit test', () => {
             rpcMock.expects("registerMethod").withExactArgs({ func: exploregens["isLegalNoteAccepted"], thisArg: exploregens });
             rpcMock.expects("registerMethod").withExactArgs({ func: exploregens["acceptLegalNote"], thisArg: exploregens });
 
-            workspaceConfigMock.expects("get").withExactArgs(ExploreGens["INSTALLATION_LOCATION"]).returns();
+            workspaceConfigMock.expects("get").withExactArgs(ExploreGens["INSTALLATION_LOCATION"]).returns("");
```
Comment (slavik-lvovsky) [inline backend/tests/exploregens.spec.ts:222]: it is better to remove .returns("") or .returns() if you do not use the returned value

**Review by slavik-lvovsky — CHANGES_REQUESTED** [review]

File: backend/tests/exploregens.spec.ts (hunk 4 above)
Comment (tomer-epstein) [reply backend/tests/exploregens.spec.ts:347]: so, in this case it 100% that there '~/notExistLocation' will not exist on test system.

**Review by slavik-lvovsky — APPROVED** [review]

**Pull request merged**